src/backend/gsroa.c

```diff
@@ -49,15 +54,66 @@ static void sliceStructs_Gather(SymInfo *sia, elem *e)
                 if (si >= 0 && sia[si].canSlice)
                 {
                     assert(si < globsym.top);
-                    if (tysize(e->Ety) != REGSIZE ||
-                        (e->Eoffset != 0 && e->Eoffset != REGSIZE))
+                    unsigned sz = tysize(e->Ety);
+                    if (sz == 2 * REGSIZE)
+                    {
+                        // Rewrite as OPpair later
+                        sia[si].usePair = true;
+
+                        /* OPpair cannot handle XMM registers, cdpair() and fixresult()
+                         */
+                        if (tyfloating(sia[si].ty0) || tyfloating(sia[si].ty1))
+                            sia[si].canSlice = false;
+                    }
+                    else if (sz == REGSIZE &&
+                        (e->Eoffset == 0 || e->Eoffset == REGSIZE))
+                    {
+                        if (!sia[si].accessSlice)
+                        {
+                            sia[si].ty0 = TYnptr;
+                            sia[si].ty1 = TYnptr;
+                        }
+                        sia[si].accessSlice = true;
+                        if (e->Eoffset == 0)
+                            sia[si].ty0 = tybasic(e->Ety);
+                        else
+                            sia[si].ty1 = tybasic(e->Ety);
+                        if (sia[si].usePair && (tyfloating(sia[si].ty0) || tyfloating(sia[si].ty1)))
```

Comment:
This is weird, when `usePair` gets set above (`if (sz == 2 * REGSIZE)`), it expects ty0/1 to already be set. But here where the ty0/1 assignment is done, you're checking for `usePair` to be set earlier.
How can both cases be valid?